[KHRGA-80] Updated DA Host allocation and memory sharing (#104)
KhronosGroup/SYCL_Reference: source/iface/host-allocation.rst

  }

When an allocator returns a ``nullptr``, the runtime
cannot allocate data on the host. Note that in this case
the runtime will raise an error if it requires host memory
but it is not available (e.g when moving data across SYCL
backend contexts).

In some cases, the implementation may retain a copy of
the allocator object even after the buffer is destroyed.
For example, this can happen when the buffer object is
destroyed before commands using accessors to the buffer
have completed. Therefore, the application must be
prepared for calls to the allocator even after the buffer
is destroyed.

.. note::

  If the application needs to know when the implementation
  has destroyed all copies of the allocator, it can maintain
  a reference count within the allocator.

The definition of allocators extends the current functionality
of SYCL, ensuring that users can define allocator functions for
specific hardware or certain complex shared memory mechanisms,
and improves interoperability with STL-based libraries.

.. seealso:: |SYCL_SPEC_HOST_ALLOC|

Default allocators
==================

A default allocator is always defined by the implementation.

> For allocations greater than size zero, it is guaranteed to

Reviewer: ```suggestion
For allocations greater than size zero, when successful it is guaranteed to
```
I've opened https://github.com/KhronosGroup/SYCL-Docs/pull/501 to fix upstream

as the one passed in the buffer constructor, but it is the
responsibility of the user to handle the potential race conditions.

The list of SYCL default allocators:

.. list-table::
  :header-rows: 1

  * - Allocator
    - Description
  * - ``template <class T> sycl::buffer_allocator``
    - It is the default :ref:`buffer` allocator used by the runtime,
      when no allocator is defined by the user.

      Meets the C++ named requirement Allocator.

      A buffer of data type ``const T`` uses ``buffer_allocator<T>``
      by default.
  * - ``sycl::image_allocator``
    - It is the default allocator used by the runtime for the SYCL :ref:`unsampled_image`
      and :ref:`sampled_image` classes when no allocator is provided by the user.

      The ``sycl::image_allocator`` is required to allocate in elements of ``std::byte``.

.. _host_memory_sharing:

=========================================================
Sharing host memory with the SYCL data management classes
=========================================================

In order to allow the SYCL runtime to do memory management
and allow for data dependencies, there are :ref:`buffer`

> and :ref:`image <iface-images>` classes defined.

Reviewer: This renders as:
> sycl::buffer and image classes defined.

We should have consistent namespace qualification, instead of one with `sycl::` qualification and the other without

Author: The idea was to unite image classes (``sycl::sampled_image`` and ``sycl::unsampled_image``) into one entity ``image`` and leave a reference to the Images page. Or I can modify `` :ref:`image <iface-images>` `` as `` :ref:`unsampled_image`/:ref:`sampled_image` `` so it will render it as ``sycl::unsampled_image/sycl::sampled_image`` but it will take more space and could be a bit confusing from my perspective.

Which approach is better?

Reviewer: OK, I understand.  I can't think of a much better solution here, so please just leave as is in the PR without change.

> The default behavior for them is that a “raw” pointer

Reviewer: ```suggestion
The default behavior for them is that if a “raw” pointer
```

is given during the construction of the data management

> class, with full ownership to use it until the destruction

Reviewer: ```suggestion
class, then full ownership to use it is given to the SYCL runtime until the destruction
```

of the SYCL object.

Below you can find details on sharing or explicitly not
sharing host memory with the SYCL data classes. The same

> rules will apply to :ref:`images <iface-images>` as well.

Reviewer: ```suggestion
rules apply to :ref:`images <iface-images>` as well.
```

.. seealso:: |SYCL_SPEC_HOST_MEM_SHARING|

Default behavior
================

> When using a :ref:`buffer`, the ownership of the pointer

Reviewer: ```suggestion
When using a :ref:`buffer`, the ownership of a pointer
```

passed to the constructor of the class is, by default,
passed to SYCL runtime, and that pointer cannot be used
on the host side until the :ref:`buffer` or
:ref:`image <iface-images>` is destroyed.

A SYCL application can access the contents of the memory
managed by a SYCL buffer by using a :ref:`host_accessor`

> as defined in. However, there is no guarantee that the

Reviewer: ```suggestion
. However, there is no guarantee that the
```

host accessor synchronizes with the original host
address used in its constructor.

The pointer passed in is the one used to copy data back
to the host, if needed, before buffer destruction.

> The memory pointed by host pointer will not be deallocated

Reviewer: ```suggestion
The memory pointed to by host pointer will not be deallocated
```

by the runtime, and the data is copied back from the device

> if there is a need for it.

Reviewer: ```suggestion
to the host through the host pointer if there is a need for it.
```

SYCL ownership of the host memory
=================================

In the case where there is host memory to be used for
initialization of data but there is no intention of using
that host memory after the :ref:`buffer` is destroyed,
then the :ref:`buffer` can take full ownership of that
host memory.

When a :ref:`buffer` owns the host pointer there is no copy back,
by default. In this situation, the SYCL application may pass a
unique pointer to the host data, which will be then used by the
runtime internally to initialize the data in the device.

For example, the following could be used:

::

  {
    auto ptr = std::make_unique<int>(-1234);
    buffer<int, 1> b { std::move(ptr), range { 1 } };
    // ptr is not valid anymore.
    // There is nowhere to copy data back
  }

However, optionally the ``sycl::buffer::set_final_data()`` can be

> set to a ``std::weak_ptr`` to enable copying data back, to another

Reviewer: ```suggestion
set to an output iterator (including a raw pointer) or to a ``std::weak_ptr`` to enable copying data back, to another
```

host memory address that is going to be valid after :ref:`buffer`

> construction.

Author: Hm, yes, I also think it should be "destruction". Fixed in the last commit.

Maybe it's better to fix it in the specs also https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_sycl_ownership_of_the_host_memory

> construction.

Reviewer: Thanks.  Discussion will happen [here](https://github.com/KhronosGroup/SYCL-Docs/pull/502), but this PR doesn't need to wait for that.
